File: apps/api/routers/model_config.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import Optional

from ..dependencies.auth import get_current_user
from ..dependencies.database import get_db
from ..models.database import User
from ..models.model_config import (
    ModelConfigCreate, 
    ModelConfigUpdate, 
    ModelConfigResponse, 
    ModelConfigListResponse,
    ActiveModelConfigResponse,
    SetActiveModelRequest
)
from ..services.model_config_service import ModelConfigService

router = APIRouter(prefix="/model-configs", tags=["model-configs"])

# 统一API响应格式
class ApiResponse:
    @staticmethod
    def success(data=None, message="操作成功"):
        return {
            "code": 200,
            "message": message,
            "data": data
        }
    
    @staticmethod
    def error(message="操作失败", code=400):
        return {
            "code": code,
            "message": message,
            "data": None
        }


@router.post("/")
async def create_model_config(
    config_data: ModelConfigCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """创建大模型配置"""
    service = ModelConfigService(db)
    
    try:
        config = service.create_config(config_data, current_user.id)
        return ApiResponse.success(config, "配置创建成功")
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"创建配置失败: {str(e)}"
        )


@router.get("/")
async def get_model_configs(
    model_type: Optional[str] = None,
    page: int = 1,
    limit: int = 100,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """获取大模型配置列表"""
    service = ModelConfigService(db)
    
    # 验证model_type参数
    if model_type and model_type not in ["llm", "vision", "deepthink"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="model_type必须是llm、vision或deepthink之一"
        )
    
    configs, total = service.get_configs(model_type, page, limit)
    
    return ApiResponse.success({
        "total": total,
        "configs": configs
    }, "获取配置列表成功")


@router.get("/active")
async def get_active_configs(
    db: Session = Depends(get_db)
):
    """获取当前激活的大模型配置（无需认证，供前端使用）"""
    service = ModelConfigService(db)
    active_configs = service.get_active_configs()
    
    return ApiResponse.success({
        "llm": active_configs["llm"],
        "vision": active_configs["vision"],
        "deepthink": active_configs["deepthink"]
    }, "获取激活配置成功")


@router.get("/{config_id}")
async def get_model_config(
    config_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """获取指定ID的大模型配置"""
    service = ModelConfigService(db)
    config = service.get_config_by_id(config_id)
    
    if not config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="配置不存在"
        )
    
    return ApiResponse.success(config, "获取配置成功")


@router.put("/{config_id}")
async def update_model_config(
    config_id: int,
    config_data: ModelConfigUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """更新大模型配置"""
    service = ModelConfigService(db)
    
    config = service.update_config(config_id, config_data)
    if not config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="配置不存在"
        )
    
    return ApiResponse.success(config, "配置更新成功")


@router.delete("/{config_id}")
async def delete_model_config(
    config_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """删除大模型配置"""
    service = ModelConfigService(db)
    
    try:
        success = service.delete_config(config_id)
        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="配置不存在"
            )
        
        return ApiResponse.success(None, "配置删除成功")
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.post("/set-active")
async def set_active_model(
    request: SetActiveModelRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """设置激活的大模型配置"""
    service = ModelConfigService(db)
    
    success = service.set_active_config(request.model_type, request.config_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="设置激活配置失败，请检查配置ID和模型类型是否匹配"
        )
    
    return ApiResponse.success(None, f"{request.model_type}模型配置已切换成功") 

File: apps/web/src/stores/modelConfig.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import type { ModelConfig, ModelConfigCreate, ModelConfigUpdate, ActiveModelConfigs } from '../types/modelConfig'
import { modelConfigApi } from '../services/api/modelConfig.api'

export const useModelConfigStore = defineStore('modelConfig', () => {
  const configs = ref<ModelConfig[]>([])
  const activeConfigs = ref<ActiveModelConfigs>({})
  const loading = ref(false)
  const error = ref<string | null>(null)

  const fetchConfigs = async (modelType?: string) => {
    loading.value = true
    error.value = null
    try {
      console.log('正在获取配置，模型类型:', modelType)
      const response = await modelConfigApi.getConfigs(modelType)
      console.log('API响应:', response)
      
      // 处理统一的API响应格式 {code, message, data}
      if (response && response.data && response.data.configs && Array.isArray(response.data.configs)) {
        configs.value = response.data.configs
        console.log('成功获取配置:', configs.value)
      } else {
        console.error('API响应数据格式不正确:', response)
        error.value = 'API响应数据格式不正确'
        configs.value = []
      }
    } catch (err) {
      // 对于401错误（未认证），这在首页是预期的，不需要记录错误
      if (err instanceof Error && (err.message.includes('401') || err.message.includes('Unauthorized'))) {
        console.info('未认证访问，使用默认配置')
        configs.value = []
        error.value = null // 清除错误状态
      } else {
        console.error('获取配置失败:', err)
        error.value = err instanceof Error ? err.message : '获取配置失败'
        configs.value = []
      }
    } finally {
      loading.value = false
    }
  }

  const fetchActiveConfigs = async () => {
    try {
      const response = await modelConfigApi.getActiveConfigs()
      console.log('获取激活配置响应:', response)
      if (response && response.data) {
        activeConfigs.value = response.data
      }
    } catch (error) {
      // 对于401错误，静默处理，保持activeConfigs为空对象
      if (error instanceof Error && (error.message.includes('401') || error.message.includes('Unauthorized'))) {
        console.info('未认证访问激活配置，将使用默认配置')
        activeConfigs.value = {}
      } else {
        console.error('获取激活配置失败:', error)
      }
    }
  }

  const fetchPublicConfigs = async (modelType: string) => {
    try {
      const response = await modelConfigApi.getPublicConfigs(modelType)
      console.log(`获取${modelType}公共配置响应:`, response)
      if (response && response.data && response.data.configs && Array.isArray(response.data.configs)) {
        // 将公共配置合并到configs中
        const publicConfigs = response.data.configs
        // 移除已存在的同类型配置，然后添加新的
        configs.value = configs.value.filter(config => config.model_type !== modelType)
        configs.value.push(...publicConfigs)
        console.log(`成功获取${modelType}公共配置:`, publicConfigs)
      }
    } catch (error) {
      console.error(`获取${modelType}公共配置失败:`, error)
    }
  }

  const createConfig = async (configData: ModelConfigCreate) => {
    try {
      const response = await modelConfigApi.createConfig(configData)
      console.log('创建配置响应:', response)
      if (response && response.data) {
        configs.value.push(response.data)
        return response.data
      }
      throw new Error('创建配置响应格式不正确')
    } catch (error) {
      console.error('创建配置失败:', error)
      throw error
    }
  }

  const updateConfig = async (id: number, configData: ModelConfigUpdate) => {
    try {
      const response = await modelConfigApi.updateConfig(id, configData)
      console.log('更新配置响应:', response)
      if (response && response.data) {
        const index = configs.value.findIndex(config => config.id === id)
        if (index !== -1) {
          configs.value[index] = response.data
        }
        return response.data
      }
      throw new Error('更新配置响应格式不正确')
    } catch (error) {
      console.error('更新配置失败:', error)
      throw error
    }
  }

  const deleteConfig = async (id: number) => {
    try {
      await modelConfigApi.deleteConfig(id)
      configs.value = configs.value.filter(config => config.id !== id)
    } catch (error) {
      console.error('删除配置失败:', error)
      throw error
    }
  }

  const setActiveConfig = async (modelType: string, configId: number) => {
    try {
      await modelConfigApi.setActiveConfig(modelType, configId)
      
      // 更新本地状态
      configs.value.forEach(config => {
        if (config.model_type === modelType) {
          config.is_active = config.id === configId
        }
      })
      
      // 重新获取激活配置
      await fetchActiveConfigs()
    } catch (error) {
      console.error('设置激活配置失败:', error)
      throw error
    }
  }

  const getActiveConfigByType = (modelType: string) => {
    return activeConfigs.value[modelType as keyof ActiveModelConfigs]
  }

  return {
    configs,
    activeConfigs,
    loading,
    error,
    fetchConfigs,
    fetchActiveConfigs,
    fetchPublicConfigs,
    createConfig,
    updateConfig,
    deleteConfig,
    setActiveConfig,
    getActiveConfigByType
  }
}) 